### DISCUSSION TITLE
Access to filer website with a link-scoped ipv6 address

### DISCUSSION THREAD

**Discussion by jiaqigan** [seaweedfs/seaweedfs#2339] [Q&A]
I started a server with command './weed -logdir=./data/log server -filer=true -dir=./volume/vol1 -ip="[fe80::d1cc:6839:faa8:c3bc%ens33]" '.
How can I access to the filer website with port 8888?
And I got this error when upload files via filer server by command 'curl -g -F "filaname=@hello.txt" "http://[fe80::d1cc:6839:faa8:c3bc%ens33]:8888/" '
{"error":"create upload request http://[fe80::d1cc:6839:faa8:c3bc%ens33]:8080/4,0982f00ab7: parse \"http://[fe80::d1cc:6839:faa8:c3bc%ens33]:8080/4,0982f00ab7\": invalid URL escape \"%en\""}

**Comment by chrislusf** [comment]
See https://github.com/chrislusf/seaweedfs/wiki/FAQ#support-ipv6

The `-ip` and `-ip.bind` have different format for IPv6.

**Reply by jiaqigan** [reply]
My ifconfig info are listed below:
ens33: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.95.129  netmask 255.255.255.0  broadcast 192.168.95.255
        inet6 fe80::1ba9:44c5:9d47:af7a  prefixlen 64  scopeid 0x20<link>

And I changed my command to: ./weed -logdir=./data/log server -filer=true -dir=./volume -ip="[fe80::1ba9:44c5:9d47:af7a]" -ip.bind="[fe80::1ba9:44c5:9d47:af7a%ens33]"

But when I tried to start the server, I still got these errors:

I0917 09:03:23  4042 master.go:166] current: [fe80::1ba9:44c5:9d47:af7a]:9333 peers:
I0917 09:03:23  4042 file_util.go:23] Folder ./volume Permission: -rwxr-xr-x
I0917 09:03:23  4042 master.go:166] current: [fe80::1ba9:44c5:9d47:af7a]:9333 peers:
I0917 09:03:23  4042 file_util.go:23] Folder ./volume Permission: -rwxr-xr-x
I0917 09:03:23  4042 master_server.go:116] Volume Size Limit is 30000 MB
I0917 09:03:23  4042 master_server.go:210] adminScripts:
I0917 09:03:23  4042 master.go:121] Start Seaweed Master 30GB 2.68 96514f0f00f1ca6f0f253c1c4080a45887d09ad9 at [fe80::1ba9:44c5:9d47:af7a%ens33]:9333
I0917 09:03:23  4042 volume_grpc_client_to_master.go:41] checkWithMaster [fe80::1ba9:44c5:9d47:af7a]:9333: get master [fe80::1ba9:44c5:9d47:af7a]:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp [fe80::1ba9:44c5:9d47:af7a]:19333: connect: invalid argument"
I0917 09:03:23  4042 raft_server.go:71] Starting RaftServer with [fe80::1ba9:44c5:9d47:af7a]:9333
I0917 09:03:23  4042 raft_server.go:131] current cluster leader: 
I0917 09:03:25  4042 volume_grpc_client_to_master.go:41] checkWithMaster [fe80::1ba9:44c5:9d47:af7a]:9333: get master [fe80::1ba9:44c5:9d47:af7a]:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp [fe80::1ba9:44c5:9d47:af7a]:19333: connect: invalid argument"
I0917 09:03:27  4042 volume_grpc_client_to_master.go:41] checkWithMaster [fe80::1ba9:44c5:9d47:af7a]:9333: get master [fe80::1ba9:44c5:9d47:af7a]:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp [fe80::1ba9:44c5:9d47:af7a]:19333: connect: invalid argument"
I0917 09:03:28  4042 volume_grpc_client_to_master.go:41] checkWithMaster [fe80::1ba9:44c5:9d47:af7a]:9333: get master [fe80::1ba9:44c5:9d47:af7a]:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp [fe80::1ba9:44c5:9d47:af7a]:19333: connect: invalid argument"
I0917 09:03:30  4042 volume_grpc_client_to_master.go:41] checkWithMaster [fe80::1ba9:44c5:9d47:af7a]:9333: get master [fe80::1ba9:44c5:9d47:af7a]:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp [fe80::1ba9:44c5:9d47:af7a]:19333: connect: invalid argument"
I0917 09:03:32  4042 volume_grpc_client_to_master.go:41] checkWithMaster [fe80::1ba9:44c5:9d47:af7a]:9333: get master [fe80::1ba9:44c5:9d47:af7a]:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp [fe80::1ba9:44c5:9d47:af7a]:19333: connect: invalid argument"
I0917 09:03:34  4042 volume_grpc_client_to_master.go:41] checkWithMaster [fe80::1ba9:44c5:9d47:af7a]:9333: get master [fe80::1ba9:44c5:9d47:af7a]:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp [fe80::1ba9:44c5:9d47:af7a]:19333: connect: invalid argument"
I0917 09:03:35  4042 volume_grpc_client_to_master.go:41] checkWithMaster [fe80::1ba9:44c5:9d47:af7a]:9333: get master [fe80::1ba9:44c5:9d47:af7a]:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp [fe80::1ba9:44c5:9d47:af7a]:19333: connect: invalid argument"
I0917 09:03:37  4042 volume_grpc_client_to_master.go:41] checkWithMaster [fe80::1ba9:44c5:9d47:af7a]:9333: get master [fe80::1ba9:44c5:9d47:af7a]:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp [fe80::1ba9:44c5:9d47:af7a]:19333: connect: invalid argument"

**Reply by chrislusf** [reply]
@nivekuil do you have any suggestions?

**Reply by nivekuil** [reply]
afaik you always have to specify the interface with a link-local ipv6. Apparently you cannot access a link local ipv6 in most browsers: 

> However browsers like Chrome or Firefox do not support entering link local addresses inside a URL, which prevents accessing devices locally with a browser, for instance for configuring them.
https://ungleich.ch/u/blog/ipv6-link-local-support-in-browsers/

The error when trying to `curl` a file seems like it would be up to seaweedfs to handle though. It seems like handling ipv6 link locals in URLs is an unsolved problem. Maybe you can point a DNS name, say in /etc/hosts, to the IP? Or assign a global IP to the interface instead of using the link local, which is probably the correct thing to do

**Reply by chrislusf** [reply]
The error was trying to use grpc to connect to `<ipV6>:19333`, which should work.

**Reply by nivekuil** [reply]
I don't think it should work since it's not a valid ipv6 address without the %device. Ping will give the same connect: invalid argument error for example https://stackoverflow.com/questions/12260003/connect-returns-invalid-argument-with-ipv6-address

My above comment is replying to the original post, not the grpc error which I think is expected.

**Reply by chrislusf** [reply]
The volume server needs to use `<ipV6>:19333` to connect to the master. Did it work for you when using ipV6 ?

**Reply by nivekuil** [reply]
I haven't tried connecting to master directly over ipv6, but my master server can reach the volume servers over ipv6 properly. I assume that @jiaqigan did not see the grpc connection error when the ip included %ens33, only when it was removed per your suggestion, which I think is not correct because link-local ipv6s cannot be used without the %

**Reply by chrislusf** [reply]
@nivekuil what are your command lines to start master and volume servers?

**Reply by nivekuil** [reply]
It's not really relevant, but master is in a consul connect service mesh, so it has -ip and -ip.bind 127.0.0.1.  The volume server talks to it through an envoy sidecar proxy. Volume does not set an ip explicitly, its pod has a global unicast ipv6 and the seaweedfs ip is just that, found automatically.
